pkg/reconciler/apis/apibinding/apibinding_permission_claim_reconcile.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apibinding
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1/permissionclaims"
+)
+
+type permissionClaim struct {
+	claim apisv1alpha1.PermissionClaim
+	label string
+	key   string
+}
+
+func (c *controller) reconcilePermissionClaims(ctx context.Context, apiBinding *apisv1alpha1.APIBinding) error {
+	// this will be responsible for making sure that the the observed state is equal to the accepted state. it should handle GC of claims from the given objects.
+
+	if len(apiBinding.Spec.AcceptedPermissionClaims) == 0 && len(apiBinding.Status.ObservedAcceptedPermissionClaims) == 0 {
+		return nil
+	}
+
+	lc := logicalcluster.From(apiBinding)
+
+	dc := c.discoveryClusterClient.WithCluster(lc)
+
+	// TODO: this is not scalable and we will need to figure out a better way.
+	// One option is an informer/cache
+	apiGroupResources, err := restmapper.GetAPIGroupResources(dc)
+	if err != nil {
+		return err
+	}
+	mapper := restmapper.NewDiscoveryRESTMapper(apiGroupResources)
+
+	// get the added permission claims if added, add are just requeue the object.
+	addedPermissionClaims := map[schema.GroupVersionResource][]permissionClaim{}
+	for _, acceptedPC := range apiBinding.Spec.AcceptedPermissionClaims {
+		found := false
+		for _, observedPC := range apiBinding.Status.ObservedAcceptedPermissionClaims {
+			if reflect.DeepEqual(acceptedPC, observedPC) {
+				found = true
+				break
+			}
+		}
+		if !found {
+			gvr, err := mapper.ResourcesFor(schema.GroupVersionResource{Group: acceptedPC.Group, Resource: acceptedPC.Resource})
+
```

Comment:
no empty line in front of error handling.

Reply:
updated